File: app/src/main/java/com/arknights/recruitmenthelper/domain/RecruitmentCalculator.kt
package com.arknights.recruitmenthelper.domain

import com.arknights.recruitmenthelper.data.Operator
import com.arknights.recruitmenthelper.data.Tag

data class TagCombo(
    val tags: List<Tag>,
    val matchedOperators: List<Operator>,
    val floorRarity: Int,
    val ceilingRarity: Int,
) {
    val isGuaranteed: Boolean get() = floorRarity >= 4
}

class RecruitmentCalculator(operators: List<Operator>) {

    private val operatorsByTag: Map<Int, Set<Operator>> =
        operators
            .flatMap { operator -> operator.tagIds.map { tagId -> tagId to operator } }
            .groupBy({ it.first }, { it.second })
            .mapValues { it.value.toSet() }

    // Arknights recruitment only honors combinations of 1-3 selected tags.
    fun evaluate(selectedTags: List<Tag>): List<TagCombo> {
        require(selectedTags.size <= 5) { "Arknights recruitment allows at most 5 selected tags" }

        return (1..MAX_COMBO_SIZE)
            .flatMap { size -> selectedTags.combinations(size) }
            .mapNotNull(::evaluateCombo)
            .sortedWith(
                compareByDescending<TagCombo> { it.floorRarity }
                    .thenByDescending { it.ceilingRarity }
                    .thenBy { it.tags.size }
            )
    }

    private fun evaluateCombo(tags: List<Tag>): TagCombo? {
        val matched = tags
            .map { operatorsByTag[it.id].orEmpty() }
            .reduce { acc, set -> acc.intersect(set) }

        if (matched.isEmpty()) return null

        return TagCombo(
            tags = tags,
            matchedOperators = matched.sortedByDescending { it.rarity },
            floorRarity = matched.minOf { it.rarity },
            ceilingRarity = matched.maxOf { it.rarity },
        )
    }

    private companion object {
        const val MAX_COMBO_SIZE = 3
    }
}

private fun <T> List<T>.combinations(size: Int): List<List<T>> {
    if (size == 0) return listOf(emptyList())
    if (size > this.size) return emptyList()
    return indices.flatMap { i ->
        drop(i + 1).combinations(size - 1).map { rest -> listOf(this[i]) + rest }
    }
}

File: windows-companion/RecruitmentCore/RecruitmentCalculator.cs
namespace RecruitmentCore;

public record TagCombo(
    IReadOnlyList<Tag> Tags,
    IReadOnlyList<Operator> MatchedOperators,
    int FloorRarity,
    int CeilingRarity)
{
    public bool IsGuaranteed => FloorRarity >= 4;
}

// Direct port of the Kotlin RecruitmentCalculator: enumerate every 1-3 tag
// subset of the selection, intersect each subset's matching operators, then
// rank by guaranteed floor rarity (highest first).
public class RecruitmentCalculator
{
    private const int MaxComboSize = 3;

    private readonly Dictionary<int, HashSet<Operator>> _operatorsByTag;

    public RecruitmentCalculator(IEnumerable<Operator> operators)
    {
        _operatorsByTag = new Dictionary<int, HashSet<Operator>>();
        foreach (var op in operators)
        {
            foreach (var tagId in op.TagIds)
            {
                if (!_operatorsByTag.TryGetValue(tagId, out var set))
                {
                    set = new HashSet<Operator>();
                    _operatorsByTag[tagId] = set;
                }
                set.Add(op);
            }
        }
    }

    public List<TagCombo> Evaluate(IReadOnlyList<Tag> selectedTags)
    {
        if (selectedTags.Count > 5)
            throw new ArgumentException("Arknights recruitment allows at most 5 selected tags");

        var combos = new List<TagCombo>();
        for (var size = 1; size <= MaxComboSize; size++)
        {
            foreach (var combo in Combinations(selectedTags, size))
            {
                var evaluated = EvaluateCombo(combo);
                if (evaluated is not null) combos.Add(evaluated);
            }
        }

        combos.Sort((a, b) =>
        {
            var cmp = b.FloorRarity.CompareTo(a.FloorRarity);
            if (cmp != 0) return cmp;
            cmp = b.CeilingRarity.CompareTo(a.CeilingRarity);
            if (cmp != 0) return cmp;
            return a.Tags.Count.CompareTo(b.Tags.Count);
        });

        return combos;
    }

    private TagCombo? EvaluateCombo(IReadOnlyList<Tag> tags)
    {
        HashSet<Operator>? matched = null;
        foreach (var tag in tags)
        {
            var set = _operatorsByTag.GetValueOrDefault(tag.Id) ?? new HashSet<Operator>();
            matched = matched is null ? new HashSet<Operator>(set) : new HashSet<Operator>(matched.Intersect(set));
        }

        if (matched is null || matched.Count == 0) return null;

        return new TagCombo(
            Tags: tags.ToList(),
            MatchedOperators: matched.OrderByDescending(o => o.Rarity).ToList(),
            FloorRarity: matched.Min(o => o.Rarity),
            CeilingRarity: matched.Max(o => o.Rarity));
    }

    private static IEnumerable<List<T>> Combinations<T>(IReadOnlyList<T> items, int size)
    {
        if (size == 0)
        {
            yield return new List<T>();
            yield break;
        }
        if (size > items.Count) yield break;

        for (var i = 0; i <= items.Count - size; i++)
        {
            var rest = items.Skip(i + 1).ToList();
            foreach (var tail in Combinations(rest, size - 1))
            {
                var combo = new List<T> { items[i] };
                combo.AddRange(tail);
                yield return combo;
            }
        }
    }
}
